### ISSUE TITLE
Compile check with multi module setup. 

### ISSUE THREAD

**Issue by extmkv** [InsertKoinIO/koin-annotations#103]
**Describe the bug**
Our project has 90+ modules, where 38 use Koin. After finishing the migration to Koin Annotations, we enabled the flag in all modules, and an issue arises when a module requires a dependency provided in another one, causing it to fail.

**To Reproduce**
Lets take into consideration a project with 3 library modules + 1 app module.

> - Module A: Domain module
> - Module B: Implements some interfaces from Module A and register on Koin
> - Module C: It's a feature an needs to get some stuff from Koin.

**Module A**
This module contains an interface `Foo`.

**Module B**
Has an implementation of `Foo` called `Bar`

```kotlin
@Single
internal fun provideFoobar() : Foo {
     return Bar()
}
```

**Module C**

```kotlin
@Factory
internal fun provideMyFactory(
    foo: Foo
) : MyFactoryObject {
     return MyFactory.build(too)
}
```

**App**
```kotlin
 startKoin {
            androidLogger(Level.ERROR)
            androidContext(this@TrilogyApp)
            modules(listOf(ModuleA().module, ModuleB().module))
        }
```

---

*Error*

```
e: [ksp] --> Missing Definition type 'com.example.Foo' for 'com.example.b.provideMyFactory'. Fix your configuration to define type 'Foo'.
```

I'm gonna create a repository with a similar setup.

**Comment by rodrirepresa** [comment]
Here is a repo to reproduce @extmkv issue:
https://github.com/rodrirepresa/koinannotationissue

**Comment by extmkv** [comment]
@arnaudgiuliani do you have any insights or idea how we can solve this?

**Comment by TheReprator** [comment]
One way is to create a dependency graph from there like we do with hilt or dagger, and remove the annotation from the class, it will work, bcz i had faced the issue, and was able to sort in this way, so i hope, it will get sorted

**Comment by extmkv** [comment]
@TheReprator can you provide an example?

**Comment by TheReprator** [comment]
You can see this repo, for this ise case,

Repository: https://github.com/TheReprator/AccountBook_Backend/tree/koinAnnotation
Branch: koinAnnotation


But you will not be able to run this, as it have a decency which I need to provide dynamically, and I had raised a bug in koin for that,

Regards,
Vikram Singh

**Comment by arnaudgiuliani** [comment]
@extmkv did you use includes to be sure you chain the modules?

C uses B uses A. Then you need to include A in B, and B in C.

**Comment by ZackOPP** [comment]
> C uses B uses A. Then you need to include A in B, and B in C.

I have the same issue. Using `include` doesn't work for me, I still get the issue about the missing dependency.

**Comment by extmkv** [comment]
@arnaudgiuliani I don't think that will work at all. Did you had time to check the repository shared by @rodrirepresa ?

**Comment by arnaudgiuliani** [comment]
Ok, need to check more deeply 👍

**Comment by extmkv** [comment]
@arnaudgiuliani did you manage to have a look?

**Comment by rubenquadros** [comment]
I think I have a similar issue. Is this even possible?
```
interface Foo {
    fun doSomething()
}

@Single
internal class Bar : Foo {
    override fun doSomething() { }
}    
```

**Comment by krzdabrowski** [comment]
@arnaudgiuliani any updates on this? This issue still occurs on KA 1.4.0/2.0.0-Beta1 between two feature modules

**Comment by BraveOwlet** [comment]
@arnaudgiuliani Is there a way to fix it?
I also have this problem on KA 1.4.0/2.0.0-Beta1

**Comment by abir-hasan** [comment]
@extmkv @rodrirepresa were you guys able to come up with a workaround?

**Comment by arnaudgiuliani** [comment]
Still stuck? Ok, I'm investigating on it 👍 
@extmkv @krzdabrowski @BraveOwlet

**Comment by arnaudgiuliani** [comment]
Temporary manner you can add `@Provided`, but I would like to avoid such. Problem is that module verification in module C will try to look at existing ref. Then it should includes module B.

**Comment by arnaudgiuliani** [comment]
TL;DR: You need classpath visiblity on your included definition. Koin can implicilty let you go. But Compile safety will need to have module included in the main module. If you try to combine modules with "+", it won't be seen as "module include". 

After rereading all your case @extmkv , you module C need to includes Module B as it needs visibility on it in the classpath. I propose to close here, we can reopen a discussion if needed.

**Comment by ArchdukeTim** [comment]
> module C need to includes Module B

What? This isn't always possible. What if B depends on C already? Does this mean that compile checks are completely unusable for anyone who has an interface in a lower-level module, but the implementation at a higher level?

**Comment by ArchdukeTim** [comment]
@arnaudgiuliani  any plans to resolve this properly?

**Comment by arnaudgiuliani** [comment]
Really weird here 🤔 Could you extract a sample project?

**Comment by TreeFrogApps** [comment]
@arnaudgiuliani Why was this closed?

I generally think what is being explained is that Koin Annotations + the compiler plugin don't work well with Clean Architecture, multi-module projects where lower-level modules use dependencies that are implemented/declared in higher-level modules. This all "wires" up in the graph inside the `:app` module. However, lower-level dependencies will fundamentally fail the build with missing definitions during their isolated compilation phase.

If my understanding is correct, this will also happen in any parent-to-child module relationship where dependency inversion is deployed, or where dependencies are declared in "higher" modules and used in "lower" modules, because the compiler plugin can only "look" in its current module and inwards/downwards to others when its doing a static analysis.

Example:

- `:domain` // lower-level dependency | core application/business rules and contracts.
- `:data` // higher-level dependency | depends on `:domain`, implements some interface contracts, and declares them with annotations like `@Single` or `@Factory`.

If `:domain` attempts to create a class that is exposed to koin and uses a interface contract as a dependency, which has a concrete implementation in `:data`, it will complain of a missing definition when it is compiled. 

```
[ :app ]          <-- Koin should check here, but it's too late.
   │
   ▼ 
[ :data ]         <-- Implementation lives here.
   │
   ▼ 
[ :domain ]       <-- ❌ Koin checks here. Cannot look up. Fails build.
```

Previously, this kind of setup with the deprecated definition DSL was easy to accomplish—it all wired up together in the main graph.

As mentioned, you can use `@Provided` or a blanket:

```
koinCompiler {
    compileSafety = false
}
```

In all modules that this is an issue, just so it compiles. However, doesn't that completely defeat the point of compile-time safety?

[The official documentation](https://insert-koin.io/docs/setup/compiler-plugin/#multi-module-projects) is slightly thin on setup details but I suspect the assumption is that multi-module setups are only ever expected to rely on dependencies inwards/downwards (apart from say whitelisted deps like Android `Context`?).

@extmkv  Is this a fair assessment and did you ever come to a solution other than architectural changes?

**Comment by arnaudgiuliani** [comment]
let's open a didicated thread for clean architecture implementation. Else title is misleading 👍
